### PR TITLE
md5: add assembly implementation for aarch64

### PR DESCRIPTION
**Pull request by AWSjswinney**
This change improves md5 performance significantly by using a hand-optimized
assembly implementation of the inner loop of md5 calculation. The instructions
are carefully ordered to separate data dependencies as much as possible.

Test with:
```
$ openssl speed md5
```

AWS Graviton 2
```
type             16 bytes    64 bytes     256 bytes    1024 bytes   8192 bytes   16384 bytes
md5              46990.60k   132778.65k   270376.96k   364718.08k   405962.75k   409201.32k
md5-modified     51725.23k   152236.22k   323469.14k   453869.57k   514102.61k   519056.04k
                 +10%        +15%         +20%         +24%         +27%         +27%
```
Apple M1
```
type             16 bytes    64 bytes     256 bytes    1024 bytes   8192 bytes   16384 bytes
md5              74634.39k   195561.25k   375434.45k   491004.23k   532361.40k   536636.48k
md5-modified     84637.11k   229017.09k   444609.62k   588069.50k   655114.24k   660850.56k
                 +13%        +17%         +18%         +20%         +23%         +23%
```

### CODE REVIEW

**Comment by t8m** [comment]
Although MD5 is still fairly widely used for non-security-relevant stuff I do not think we want to optimize its implementation as it is a seriously insecure and legacy algorithm.

**Comment by beldmit** [comment]
I'm not sure the arguments about insecurity is relevant. If we have MD5 in our distribution, it's worth having it as fast as possible.

**Comment by mattcaswell** [comment]
> I'm not sure the arguments about insecurity is relevant. If we have MD5 in our distribution, it's worth having it as fast as possible.

I think that debate is worthy of having with OTC. I've added an OTC hold.

**Comment by AGSaidi** [comment]
At least I'd appreciate having md5 go as fast as possible, people still frequently used OpenSSL's md5 implementation.

**Comment by t8m** [comment]
The problem is maintainability.

**Comment by beldmit** [comment]
Fair point.

**Comment by AWSjswinney** [comment]
I noticed I had a mistake in the description: I swapped the modified/baseline benchmark numbers for Graviton 2. I have corrected that. 


> The problem is maintainability.

@t8m I first started working on an implementation more similar to the x86_64 version which used the perl script to emit each step of each round, but I was unable to get that approach to match the performance of this one, which is more hand optimized.

**Comment by kroeckx** [comment]
I think we need an implementation that uses the perl wrappers to emit the proper assembler depending on the platform. It should end up generating the same assembler.

**Comment by kroeckx** [comment]
I'm probably confused about what's needed on arm

File: crypto/md5/asm/md5-aarch64.pl
Comment (paulidale) [inline crypto/md5/asm/md5-aarch64.pl:32]: Does this still with on armv8 without crypto?

**Review by paulidale — COMMENTED** [review]
Our usual assembly code approach checks the feature set of the CPU being used and switches between implementations.  I don't know enough about aarch64 to know if this is required here or not, however we were recently stung by ARM 64 assembler recently not being sufficiently portable between different devices.

File: crypto/md5/asm/md5-aarch64.pl
Comment (AWSjswinney) [reply crypto/md5/asm/md5-aarch64.pl:32]: Yes, it should. It's not using any special instructions. I must have left this in by mistake.

**Comment by AWSjswinney** [comment]
> I think we need an implementation that uses the perl wrappers to emit the proper assembler depending on the platform. It should end up generating the same assembler.

Yes, I agree and that's the approach that I began working with. However, another colleague of mine found an approach more complicated than would be appropriate in a perl script that yielded much better results. This approach is running 5% faster than emitting each step of each round via a perl script.

```
type             16 bytes     64 bytes    256 bytes   1024 bytes    8192 bytes   16384 bytes
md5-perl-script  50789.19k   147996.54k   312031.83k   433552.38k   488871.25k   493398.70k
md5-raw-asm      51725.23k   152236.22k   323469.14k   453869.57k   514102.61k   519056.04k
```

I would argue that maintainability for this segment of code is less important than speed, because any future work on it, is likely to completely replace this implementation again, just as this change proposes.

If it were my choice, I would choose this implementation. What would you like to do?

**Comment by paulidale** [comment]
I think the faster implementation is the way to go.

I'm more concerned about differences between different assemblers than the quality of the unrolled code.  The Perl wrappers seem to cope with multiple assemblers quite well.  Note: this might not be a problem for aarch64 using the basic instruction set, it's just appeared for lots of other platforms.

**Comment by t8m** [comment]
**Summary of the issue for OTC to decide:**

Do we want to accept (in the master branch) new optimized assembler implementations for algorithms that are legacy? At some point there was a discussion about removing assembler implementations for legacy algorithms to ease maintenance.

**Comment by paulidale** [comment]
OTC: okay for this improved assembler in master, not 3.0.

**Comment by AWSjswinney** [comment]
S3 uses md5 for content verification so md5 is still important to optimize as much a possible. I would like to see this merged into 3.0 or older branches, if possible.

**Comment by t8m** [comment]
> S3 uses md5 for content verification so md5 is still important to optimize as much a possible. I would like to see this merged into 3.0 or older branches, if possible.

No, that would require explicit approval from OMC as a new optimization is clearly a feature and not a bug fix.

**Comment by AWSjswinney** [comment]
What can I do to help get this pull request merged?

**Comment by paulidale** [comment]
It needs to be reviewed.  @openssl/committers do this.

**Review by levitte — CHANGES_REQUESTED** [review]
Readability: Assembler statements should be indented

**Comment by paulidale** [comment]
Comments in the assembly will help the next person maintain this code.  Please add some.

**Comment by AWSjswinney** [comment]
I have added comments and indented the assembly. Let me know what else you would like me to do.

Thanks!

**Review by paulidale — APPROVED** [review]
Thanks.

**Comment by paulidale** [comment]
Oops, only one review so another is required.

The tests failing is a script problem rather than what's happening.

**Comment by AWSjswinney** [comment]
@levitte I'm not sure how to clear the "change requested" flag that you added. Is that something you will have to do yourself?

**Comment by kaduk** [comment]
If you expand the "1 change requested" line at the bottom, it shows the specific review.  There should also be a three-dots menu on the right side of that line, with an option to "dismiss review", though I'm not 100% sure what permissions are needed for that to be accessible.

**Comment by AWSjswinney** [comment]
> If you expand the "1 change requested" line at the bottom, it shows the specific review. There should also be a three-dots menu on the right side of that line, with an option to "dismiss review", though I'm not 100% sure what permissions are needed for that to be accessible.

I don't seem to have the permission required to do that...

**Comment by kaduk** [comment]
> > If you expand the "1 change requested" line at the bottom, it shows the specific review. There should also be a three-dots menu on the right side of that line, with an option to "dismiss review", though I'm not 100% sure what permissions are needed for that to be accessible.
> 
> I don't seem to have the permission required to do that...

Okay, I went ahead and did it, then.

**Comment by paulidale** [comment]
We typically don't worry about the change request note.  At least if the change has been made.

**Comment by t8m** [comment]
We would like to see an external review of the assembler code by someone well versed with Aarch64 assembler.

**Comment by AWSjswinney** [comment]
@sebpop would you take a look?

**Comment by AWSjswinney** [comment]
Thanks for the contributions. I tested again and everything still passes the accuracy check and performance is improvement, though marginally. It's fewer instructions too, which is likely to help older chips a bit more.

**Review by paulidale — APPROVED** [review]

**Comment by sebpop** [comment]
I pushed some more changes to https://github.com/sebpop/openssl/tree/md5-aarch64-performance-1 
- fold lsr into add instructions (instead of saving the intermediate results in registers)
- remove register spills from loop for state->[A,B,C,D]
- reduce the number of saved registers.

Jonathan, could you please go through these changes, test, and check performance?

**Comment by AWSjswinney** [comment]
@sebpop I actually see a pretty significant regression from that change. I think it's primarily from the combined `add` and `lsr` instructions. In my testing, it's faster to issue those instructions separately than the combined form.

```
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
md5              47368.06k   139856.51k   302431.40k   428724.57k   488275.97k   493426.01k
```

**Comment by sebpop** [comment]
Thanks Jonathan for testing.
I think the last version of the patch looks good as it is.

**Comment by AWSjswinney** [comment]
@levitte Are you available to review this change?

**Comment by paulidale** [comment]
@openssl/committers review required.  The assembly should have been covered by @sebpop already but a glance over wouldn't hurt.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review

**Comment by zorrorffm** [comment]
The assembly code looks good to me except of minor x16, x17 issue.

**Comment by t8m** [comment]
> The assembly code looks good to me except of minor x16, x17 issue.

What is the issue concretely?

**Comment by AWSjswinney** [comment]
> > > The assembly code looks good to me except of minor x16, x17 issue.
> > 
> > 
> > What is the issue concretely?
> 
> As I commented in md5-aarch64.pl, it's not an issue in strict sense, just a redundant operation on x16 and x17. It's unnecessary to preserve and restore x16 and x17.

Thanks for pointing that out. I have removed x16 and x17 from the stack saved registers.

**Comment by zorrorffm** [comment]
> > > > The assembly code looks good to me except of minor x16, x17 issue.
> > > 
> > > 
> > > What is the issue concretely?
> > 
> > 
> > As I commented in md5-aarch64.pl, it's not an issue in strict sense, just a redundant operation on x16 and x17. It's unnecessary to preserve and restore x16 and x17.
> 
> Thanks for pointing that out. I have removed x16 and x17 from the stack saved registers.

Thanks!  The assembly code looks good to me.

**Review by paulidale — APPROVED** [review]
Still good, pending second review.

**Review by mattcaswell — APPROVED** [review]
LGTM. Assembly is not my thing, and I'm not familiar with aarch64, but I did spend some time looking at this code. I was able to do so because the level of commenting in this is excellent and it gave me a fighting chance of actually understanding what is going on. I wish all our assembly was like this.

**Comment by tom-cosgrove-arm** [comment]
@AGSaidi One of the team should merge this soon, just be patient - it can take a day or two

**Comment by paulidale** [comment]
Merged to master.

Merging happens when one of us notices the approve delay has passed without comments :)

**Comment by DDvO** [comment]
>>> @paulidale what is needed to merge the change at this point?
>>
>> @AGSaidi One of the team should merge this soon, just be patient - it can take a day or two
> 
> Merging happens when one of us notices the approve delay has passed without comments :)

When during 24 hours after **approval: done** was set no further change or comment was given, 
the bot @openssl-machine replaces it by **approval: ready to merge** and sends an email to the team saying
```
This pull request is ready to merge
```
This usually triggers some of us to do the merge 
but sometimes gets unnoticed, and then it's good to get a further reminder :)

**Pull request closed without merging**
